sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/CommentKeywordBase.cs

```diff
@@ -0,0 +1,122 @@
+{
+    public abstract class CommentKeywordBase : SonarDiagnosticAnalyzer
+    {
+        private const string TODO_Keyword = "TODO";
+        protected const string TODO_DiagnosticId = "S1135";
+        protected const string TODO_MessageFormat = "Complete the task associated to this '" + TODO_Keyword + "' comment.";
+
+        private const string FIXME_Keyword = "FIXME";
+        protected const string FIXME_DiagnosticId = "S1134";
+        protected const string FIXME_MessageFormat = "Take the required action to fix the issue indicated by this '" + FIXME_Keyword + "' comment.";
+
+        protected abstract DiagnosticDescriptor TODO_Diagnostic { get; }
+        protected abstract DiagnosticDescriptor FIXME_Diagnostic { get; }
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        protected CommentKeywordBase()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(TODO_Diagnostic, FIXME_Diagnostic);
+        }
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxTreeActionInNonGenerated(
+                GeneratedCodeRecognizer,
+                c =>
+                {
+                    var comments = c.Tree.GetRoot().DescendantTrivia()
+                        .Where(trivia => IsComment(trivia));
+
+                    foreach (var comment in comments)
+                    {
+                        foreach (var location in GetKeywordLocations(c.Tree, comment, TODO_Keyword))
+                        {
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(TODO_Diagnostic, location));
+                        }
+
+                        foreach (var location in GetKeywordLocations(c.Tree, comment, FIXME_Keyword))
+                        {
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(FIXME_Diagnostic, location));
+                        }
+                    }
+                });
+        }
+
+        protected abstract bool IsComment(SyntaxTrivia trivia);
+
+        private static IEnumerable<Location> GetKeywordLocations(SyntaxTree tree, SyntaxTrivia comment, string word)
+        {
+            var text = comment.ToString();
+
+            return AllIndexesOf(text, word)
+                .Where(i => IsWordAt(text, i, word.Length))
+                .Select(
+                    i =>
+                    {
+                        var startLocation = comment.SpanStart + i;
+                        var location = Location.Create(tree, TextSpan.FromBounds(startLocation, startLocation + word.Length));
+
+                        return location;
+                    });
+        }
+
+        private static IEnumerable<int> AllIndexesOf(string text, string value, StringComparison comparisonType = StringComparison.OrdinalIgnoreCase)
```

Comment:
This logic was already there before, it was just moved.

NB: I extracted the comparison type as a parameter with a default value.